### PR TITLE
Extend concord errors

### PR DESCRIPTION
**Pull request by vogre**
Add names to errors for pretty-printing
Wrap all external errors with RequestError

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "concord",
-  "version": "0.1.0",
+  "version": "0.2.0",
```
Comment (bergundy) [inline package.json]: 0.1.1?

File: src/lib/common.ts [hunk 2]
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
 export class ValidationError extends Error {
   constructor(message: string, public errors: any) {
     super(message);
+    this.name = 'ValidationError';
```
Comment (bergundy) [inline src/lib/common.ts]: Or just:
```typescript
class ValidationError extends Error {
  public readonly name = 'ValidationError';
  ...
}
```

File: src/lib/common.ts [hunk 3]
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
...
   }
 }
 
 export interface ClassValidator {
   [method: string]: Ajv.ValidateFunction;
 }
 
+export class RequestError extends Error {
+  public cause: any;
```
Comment (bergundy) [inline src/lib/common.ts]: `name` can be hardcoded here.

File: src/lib/common.ts [hunk 4]
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
...
+  public method: string;
+  public options: any;
+  constructor(message: string, cause: any, method: string, options: any) {
```
Comment (bergundy) [inline src/lib/common.ts]: Why not initialise these in the params: `public readonly cause: any`?

File: src/lib/common.ts (hunk 3 above)
Comment (bergundy) [inline src/lib/common.ts]: Prefer readonly attributes.

File: src/test/test_rpc.ts [hunk 5]
```diff
@@ -766,7 +766,17 @@ await new Promise((resolve, reject) => {
 const { address, port } = (server.address() as AddressInfo);
 const client = new TestClient('http://' + address + ':' + port);
 server.close();
-await expect(client.bar('heh')).to.eventually.be.rejectedWith(/connect ECONNREFUSED/);
+try {
+  await client.bar('heh');
```
Comment (bergundy) [inline src/test/test_rpc.ts]: You can use `.to.eventually.be.rejectedWith`

File: src/test/test_rpc.ts [hunk 6]
```diff
@@ -814,7 +824,17 @@ await new Promise((resolve, reject) => {
 });
 const { address, port } = (server.address() as AddressInfo);
 const client = new TestClient('http://' + address + ':' + port);
-await expect(client.bar('heh')).to.eventually.be.rejectedWith(Error, '500 - Internal Server Error');
+try {
```
Comment (bergundy) [inline src/test/test_rpc.ts]: See above.
If `rejectedWith` is not good enough you can use:
```
const err = await client.bar('heh').catch((err) => err);
```

File: src/test/test_rpc.ts [hunk 7]
```diff
@@ -871,7 +899,7 @@ import fetch from 'node-fetch';
 import { TestClient } from './client';
 
 export default async function test(client: TestClient) {
-  const url = (client as any /* access protected member */).serverUrl;
+  const url = client.serverUrl;
```
Comment (bergundy) [inline src/test/test_rpc.ts:885]: That makes sense..

File: templates/ts/client-browser.ts [hunk 8]
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
...
+        },
+        body: JSON.stringify(body),
+        method: 'POST',
+      });
+      isJSON = (response.headers.get('content-type') || '').startsWith('application/json');
+      if (isJSON && (response.status >= 200 && response.status < 300 || response.status === 400 || response.status === 500)) {
+        responseBody = await response.json();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, "{{name}}", { serverUrl: this.serverUrl, ...this.options, ...options });
```
Comment (bergundy) [inline templates/ts/client-browser.ts]: Can we use single quotes for `"{{name}}"`?

Comment (bergundy) [inline templates/ts/client-browser.ts]: You can use mergedOptions

File: templates/ts/client-browser.ts [hunk 9]
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
...
+        ...mergedOptions,
+        headers: {
+          ...mergedOptions.headers,
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(body),
+        method: 'POST',
+      });
+      isJSON = (response.headers.get('content-type') || '').startsWith('application/json');
+      if (isJSON && (response.status >= 200 && response.status < 300 || response.status === 400 || response.status === 500)) {
```
Comment (bergundy) [inline templates/ts/client-browser.ts]: Is it necessary to check the response status code?

File: templates/ts/client-browser.ts [hunk 10]
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
...
+        responseBody = await response.json();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, "{{name}}", { serverUrl: this.serverUrl, ...this.options, ...options });
+    }
     if (response.status >= 200 && response.status < 300) {
       const validator = this.validators.{{{name}}};
-      const wrapped = { returns: isJSON ? await response.json() : undefined }; // wrapped for coersion
+      const wrapped = { returns: isJSON ? responseBody : undefined }; // wrapped for coersion
```
Comment (bergundy) [inline templates/ts/client-browser.ts]: No need to check if `isJSON` anymore

File: templates/ts/client-browser.ts [hunk 11]
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
...
       return wrapped.returns as {{{returnType}}};
     } else if (!isJSON) {
-      throw new Error(`${response.status} - ${response.statusText}`);
+      // fall through to throw
     } else if (response.status === 400) {
-      const body = await response.json();
-      if (body.name === 'ValidationError') {
-        throw new ValidationError(body.message, body.errors);
+      if (responseBody.name === 'ValidationError') {
+        throw new ValidationError(responseBody.message, responseBody.errors);
```
Comment (bergundy) [inline templates/ts/client-browser.ts:141]: Should we double check that responseBody is an object and has `message` and `errors`?

File: templates/ts/client-node.ts [hunk 12]
```diff
@@ -1,8 +1,8 @@
 // tslint:disable
 import fetch from 'node-fetch';
-import { RequestInit } from 'node-fetch';
+import { RequestInit, Response } from 'node-fetch';
```
Comment (bergundy) [inline templates/ts/client-node.ts:3]: Not reviewing this file, assuming same diff as the browser code.

File: templates/ts/interfaces.ts [hunk 13]
```diff
@@ -2,10 +2,18 @@
 export const schema = {{{schema}}};
 
 export class InternalServerError extends Error {
+  constructor(message: string) {
```
Comment (bergundy) [inline templates/ts/interfaces.ts]: I'd put the name as `public readonly name = 'InternalServerError'`.

**Review by bergundy — COMMENTED** [review]
Overall LGTM.
nit: please use single quotes for all strings.

File: package.json (hunk 1 above)
Comment (vogre) [reply package.json]: No preference really, since all interfaces need to be regenerated anyway.

File: src/lib/common.ts (hunk 3 above)
Comment (vogre) [reply src/lib/common.ts]: Cool trick, thanks!

File: src/test/test_rpc.ts (hunk 6 above)
Comment (vogre) [reply src/test/test_rpc.ts]: This doesn't really work if there's no error.

File: src/test/test_rpc.ts (hunk 5 above)
Comment (vogre) [reply src/test/test_rpc.ts]: I can't test the error object this way.

File: templates/ts/client-browser.ts (hunk 9 above)
Comment (vogre) [reply templates/ts/client-browser.ts]: Added comment.

File: templates/ts/client-browser.ts (hunk 8 above)
Comment (vogre) [reply templates/ts/client-browser.ts]: Most likely will have a signal object which would look bad in a log line.

File: templates/ts/client-browser.ts (hunk 11 above)
Comment (vogre) [reply templates/ts/client-browser.ts:141]: In general we can probably JSON schema check the error responses, but I don't think it's likely we'll see a problem with them.

File: templates/ts/client-node.ts (hunk 12 above)
Comment (vogre) [reply templates/ts/client-node.ts:3]: Verified by diff.

**Review by vogre — COMMENTED** [review]
Thanks!

File: src/lib/common.ts [hunk 14]
```diff
@@ -11,6 +12,18 @@ export interface ClassValidator {
   [method: string]: Ajv.ValidateFunction;
 }
 
+export class RequestError extends Error {
+  public readonly name = 'RequestError';
+  constructor(
+    message: string,
+    public readonly cause: any,
```
Comment (bergundy) [inline src/lib/common.ts:24]: Could you please document what cause is?

File: src/test/test_rpc.ts (hunk 5 above)
Comment (bergundy) [reply src/test/test_rpc.ts]: You can, just `const err = await expect(client.bar('heh')).to.eventually.be.rejected`.
Please do, adding `expect(false)` is more error prone, too easy to miss.

File: templates/ts/client-browser.ts (hunk 9 above)
Comment (bergundy) [reply templates/ts/client-browser.ts]: I'm not sure we should limit parsing the response to these error codes, it makes this line too dependant on the other conditions and likely to break.

File: templates/ts/client-browser.ts (hunk 11 above)
Comment (bergundy) [reply templates/ts/client-browser.ts:141]: Agree

**Review by bergundy — COMMENTED** [review]
Thanks!

File: src/test/test_rpc.ts (hunk 5 above)
Comment (vogre) [reply src/test/test_rpc.ts]: Thanks!

It was not obvious from chai documentation...

File: templates/ts/client-browser.ts [hunk 15]
```diff
@@ -80,40 +80,54 @@ export class {{name}}Client {
...
+        method: 'POST',
+      });
+      isJSON = (response.headers.get('content-type') || '').startsWith('application/json');
+      if (isJSON) {
+        responseBody = await response.json();
+      } else {
+        responseText = await response.text();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, '{{name}}', { serverUrl: this.serverUrl, ...this.options, ...options });
```
Comment (bergundy) [inline templates/ts/client-browser.ts]: Maybe use `mergedOptions`?

**Review by bergundy — APPROVED** [review]
Thanks!

**Comment by vogre** [comment]
Thanks!

PTAL, I brought timeout code to browser generated code, it seems to be passing tests.

**Review by bergundy — APPROVED** [review]
Thanks!

**Pull request merged**